Bug: File access errors when recording Maven-run tests on Windows

Spring petclinic project. Tests pass and get recorded fine with 1.1.2, fail with 1.2.0.

Windows 10, Oracle JDK 11

```log
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.51 s <<< FAILURE! - in org.springframework.samples.petclinic.model.ValidatorTests
[ERROR] shouldNotValidateWhenFirstNameEmpty  Time elapsed: 2.5 s  <<< ERROR!
java.lang.RuntimeException: 
java.nio.file.FileSystemException: C:\Users\petrd\AppData\Local\Temp\15604754668976555540.appmap.json -> target\appmap\org_springframework_samples_petclinic_model_ValidatorTests_shouldNotValidateWhenFirstNameEmpty.appmap.json: The process cannot access the file because it is being used by another process.

	at org.springframework.samples.petclinic.model.ValidatorTests.shouldNotValidateWhenFirstNameEmpty(ValidatorTests.java:58)
Caused by: java.nio.file.FileSystemException: 
C:\Users\petrd\AppData\Local\Temp\15604754668976555540.appmap.json -> target\appmap\org_springframework_samples_petclinic_model_ValidatorTests_shouldNotValidateWhenFirstNameEmpty.appmap.json: The process cannot access the file because it is being used by another process.

	at org.springframework.samples.petclinic.model.ValidatorTests.shouldNotValidateWhenFirstNameEmpty(ValidatorTests.java:58)

```

[spring-petclinic-win-1.1.2.log](https://github.com/applandinc/appmap-java/files/6983464/spring-petclinic-win-1.1.2.log)
[spring-petclinic-win-1.2.0.log](https://github.com/applandinc/appmap-java/files/6983465/spring-petclinic-win-1.2.0.log)

Please use -e switch as described:

> To see the full stack trace of the errors, re-run Maven with the -e
switch.


On Fri, Aug 13, 2021 at 11:29 AM Petr ***@***.***> wrote:

> Spring petclinic project. Tests pass and get recorded fine with 1.1.2,
> fail with 1.2.0.
>
> Windows 10, Oracle JDK 11
>
> [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.51 s <<< FAILURE! - in org.springframework.samples.petclinic.model.ValidatorTests
> [ERROR] shouldNotValidateWhenFirstNameEmpty  Time elapsed: 2.5 s  <<< ERROR!
> java.lang.RuntimeException:
> java.nio.file.FileSystemException: C:\Users\petrd\AppData\Local\Temp\15604754668976555540.appmap.json -> target\appmap\org_springframework_samples_petclinic_model_ValidatorTests_shouldNotValidateWhenFirstNameEmpty.appmap.json: The process cannot access the file because it is being used by another process.
>
> 	at org.springframework.samples.petclinic.model.ValidatorTests.shouldNotValidateWhenFirstNameEmpty(ValidatorTests.java:58)
> Caused by: java.nio.file.FileSystemException:
> C:\Users\petrd\AppData\Local\Temp\15604754668976555540.appmap.json -> target\appmap\org_springframework_samples_petclinic_model_ValidatorTests_shouldNotValidateWhenFirstNameEmpty.appmap.json: The process cannot access the file because it is being used by another process.
>
> 	at org.springframework.samples.petclinic.model.ValidatorTests.shouldNotValidateWhenFirstNameEmpty(ValidatorTests.java:58)
>
>
> spring-petclinic-win-1.1.2.log
> <https://github.com/applandinc/appmap-java/files/6983464/spring-petclinic-win-1.1.2.log>
> spring-petclinic-win-1.2.0.log
> <https://github.com/applandinc/appmap-java/files/6983465/spring-petclinic-win-1.2.0.log>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/applandinc/appmap-java/issues/98>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVC6Z7COSWXGD4XFD4XT3T4U25TANCNFSM5CDZ3Y5A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

--- a/src/main/java/com/appland/appmap/record/Recording.java
+++ b/src/main/java/com/appland/appmap/record/Recording.java
@@ -4,45 +4,98 @@ import com.appland.appmap.util.Logger;
 
 import java.io.*;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
 
 public class Recording {
-  private final File file;
+    private final File file;
 
-  public Recording(File file) {
-    this.file = file;
-  }
-
-  public void delete() {
-    this.file.delete();
-  }
-
-  public void moveTo(String filePath) {
-    Logger.printf("Moving %s to %s\n", this.file.getPath(), filePath);
-
-    try {
-      Files.move(Paths.get(this.file.getPath()), Paths.get(filePath), StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+    public Recording(File file) {
+        this.file = file;
     }
-  }
 
-  public void readFully(boolean delete, Writer writer) throws IOException {
-    final Reader reader = new FileReader(this.file);
-    char[] buffer = new char[2048];
-    int bytesRead;
-    while ( (bytesRead = reader.read(buffer) ) != -1 ) {
-      writer.write(buffer, 0, bytesRead);
+    public void delete() {
+        this.file.delete();
     }
-    writer.flush();
 
-    if ( delete ) {
-      this.delete();
+    public void moveTo(String filePath) {
+        Path sourcePath = Paths.get(this.file.getPath());
+        Path targetPath = Paths.get(filePath);
+
+        Logger.printf("Moving %s to %s\n", sourcePath, targetPath);
+
+        Function<FileMover, FileMover.Result> tryMove = mover -> {
+            IOException exception = null;
+            try {
+                mover.move();
+            } catch (IOException e) {
+                exception = e;
+            }
+            return new FileMover.Result(exception);
+        };
+
+        FileMover[] movers = new FileMover[]{
+            () -> Files.move(sourcePath, targetPath, REPLACE_EXISTING, ATOMIC_MOVE),
+            () -> Files.move(sourcePath, targetPath, REPLACE_EXISTING),
+            () -> {
+                Files.copy(sourcePath, targetPath, REPLACE_EXISTING);
+                sourcePath.toFile().delete();
+            },
+        };
+        List<String> errors = new ArrayList<>();
+        for (FileMover mover: movers) {
+            FileMover.Result r = tryMove.apply(mover);
+            if ( r.isSucceeded() ) {
+                errors.clear();;
+                break;
+            }
+            errors.add(r.exception.getMessage());
+        }
+        if (!errors.isEmpty()) {
+            throw new RuntimeException(String.join(", ", errors));
+        }
     }
-  }
 
-  public int size() {
-    return (int) this.file.length();
-  }
+    public void readFully(boolean delete, Writer writer) throws IOException {
+        final Reader reader = new FileReader(this.file);
+        char[] buffer = new char[2048];
+        int bytesRead;
+        while ((bytesRead = reader.read(buffer)) != -1) {
+            writer.write(buffer, 0, bytesRead);
+        }
+        writer.flush();
+
+        if (delete) {
+            this.delete();
+        }
+    }
+
+    public int size() {
+        return (int) this.file.length();
+    }
+
+    interface FileMover {
+        class Result {
+            final IOException exception;
+
+            Result(IOException exception) {
+                this.exception = exception;
+            }
+
+            boolean isSucceeded() {
+                return this.exception == null;
+            }
+        }
+
+        void move() throws IOException;
+    }
 }

--- a/src/test/java/com/appland/appmap/record/RecorderTest.java
+++ b/src/test/java/com/appland/appmap/record/RecorderTest.java
@@ -16,6 +16,8 @@ import java.nio.file.Path;
 import java.util.Map;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertNotNull;
 
 public class RecorderTest {
 
@@ -73,6 +75,20 @@ public class RecorderTest {
     InputStream is = new FileInputStream(targetPath.toString());
     Map appmap = JSON.parseObject(is, Map.class);
     assertEquals("[classMap, metadata, version, events]", appmap.keySet().toString());
+  }
+
+  @Test
+  public void testUnwriteableOutputFile() throws IOException {
+    Recorder recorder = recordEvents();
+    final Recording recording = recorder.stop();
+    Exception exception = null;
+    try {
+      recording.moveTo("/no-such-directory");
+    } catch (RuntimeException e) {
+      exception = e;
+    }
+    assertNotNull(exception);
+    assertTrue(exception.toString().indexOf("java.lang.RuntimeException: ") == 0);
   }
 
   @Test

